File: frontend/src/types/goal.ts
export type GoalType = "career" | "life" | "health" | "learning" | "financial";
export type GoalStatus = "active" | "archived" | "completed";

export interface Milestone {
  id: number;
  goal_id: number;
  title: string;
  target_date: string | null;
  done: boolean;
}

export interface Goal {
  id: number;
  title: string;
  type: GoalType;
  description: string | null;
  target_date: string | null;
  status: GoalStatus;
  external_key: string | null;
  list_name: string | null;
  parent_list_name: string | null;
  created_at: string;
  updated_at: string;
  progress_pct: number;
  milestones: Milestone[];
  priority_rank?: number | null;
}

export interface GoalCreate {
  title: string;
  type: GoalType;
  description?: string;
  target_date?: string;
  list_name?: string | null;
  parent_list_name?: string | null;
}

export interface GoalUpdate {
  title?: string;
  type?: GoalType;
  description?: string;
  target_date?: string;
  status?: GoalStatus;
  list_name?: string | null;
  parent_list_name?: string | null;
}

export interface MilestoneCreate {
  title: string;
  target_date?: string;
}

export interface IngestEntityDiff {
  external_key: string;
  title: string;
  action: "create" | "update";
}

export interface IngestPreviewResult {
  goals: IngestEntityDiff[];
  tasks: IngestEntityDiff[];
  routines: IngestEntityDiff[];
  habits: IngestEntityDiff[];
}

export interface AdvisoryFieldChange {
  field: string;
  old: unknown;
  new: unknown;
}

export interface AdvisoryEntityDiff {
  external_key: string;
  title: string;
  action: "update" | "create";
  rationale: string;
  fields: AdvisoryFieldChange[];
}

export interface AdvisoryPreviewResult {
  goals: AdvisoryEntityDiff[];
  milestones: AdvisoryEntityDiff[];
  new_tasks: AdvisoryEntityDiff[];
  notes: string | null;
  session_id: string;
  generated_at: string;
}

export interface AdvisoryResult {
  created: Record<string, number>;
  updated: Record<string, number>;
  advisory_id: string;
  replayed: boolean;
}

File: backend/app/schemas/goal.py
from datetime import date, datetime
from pydantic import BaseModel
from app.models.goal import GoalType, GoalStatus


class MilestoneCreate(BaseModel):
    title: str
    target_date: date | None = None


class MilestoneUpdate(BaseModel):
    title: str | None = None
    target_date: date | None = None
    done: bool | None = None


class MilestoneRead(BaseModel):
    id: int
    goal_id: int
    title: str
    target_date: date | None
    done: bool

    model_config = {"from_attributes": True}


class GoalCreate(BaseModel):
    title: str
    type: GoalType
    description: str | None = None
    target_date: date | None = None
    external_key: str | None = None
    list_name: str | None = None
    parent_list_name: str | None = None


class GoalUpdate(BaseModel):
    title: str | None = None
    type: GoalType | None = None
    description: str | None = None
    target_date: date | None = None
    status: GoalStatus | None = None
    list_name: str | None = None
    parent_list_name: str | None = None


class GoalRead(BaseModel):
    id: int
    title: str
    type: GoalType
    description: str | None
    target_date: date | None
    status: GoalStatus
    external_key: str | None
    list_name: str | None
    parent_list_name: str | None
    created_at: datetime
    updated_at: datetime
    progress_pct: int
    milestones: list[MilestoneRead] = []

    model_config = {"from_attributes": True}
